libraries/ms-common/src/x86/critical_section.c

```diff
@@ -7,22 +7,18 @@
 #include <pthread.h>
 
 #include "interrupt_def.h"
+#include "x86_interrupt.h"
 
 static bool s_interrupts_disabled = false;
 static pthread_mutex_t s_mutex = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
 
 bool critical_section_start(void) {
   pthread_mutex_lock(&s_mutex);
   if (!s_interrupts_disabled) {
+    // Update the signal mask to prevent interrupts from being executed on the signal handler
```

Comment:
I think we should be disabling the signals before the mutex lock - our failure case right now is a signal handler attempting to lock during a critical section mutex lock.

Is there a way of detecting that we're in a signal handler? Maybe we should just not attempt to lock the mutex if we are?

Reply:
Yeah we could set a flag on entry into the signal handler and provide an API to check it (I'll do that). The issue with unmasking the signals before the mutex is that a lot of data races start occurring inside of `can_hw` I ran it with TSAN and got a ton of errors and the tests would never pass reliably. As it is I've been unable to reproduce a deadlock on the version that is my last commit regardless of how much I cap my CPU speed.

Reply:
Done.